{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c9d75a6b847760cb",
   "metadata": {
    "collapsed": false,
    "id": "c9d75a6b847760cb"
   },
   "source": [
    "# Exercise 02\n",
    "\n",
    "## Aim of the exercise:\n",
    "\n",
    "We will learn to:\n",
    "\n",
    "- use Alphafold to predict the structure of proteins\n",
    "- get simple motifs by using different protein sequences\n",
    "- learn about the Alphafold parameters\n",
    "- perform a disruption of your protein of interest\n",
    "\n",
    "This exercise is based on the work of [Sergey Ovchinnikov](https://github.com/sokrypton)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "-_lqREjHhhcT",
   "metadata": {
    "id": "-_lqREjHhhcT"
   },
   "source": [
    "# Alphafold\n",
    "\n",
    "[AlphaFold](https://deepmind.google/technologies/alphafold/), an AI system developed by DeepMind, has solved the complex protein-folding problem, allowing for almost instant and highly accurate predictions of protein structures, which are crucial for understanding cellular functions and advancing medical research. Recognized by the [Critical Assessment of protein Structure Prediction community](https://www.predictioncenter.org/), AlphaFold has [significantly](https://www.predictioncenter.org/casp14/zscores_final.cgi) expanded the availability of protein structure data through the freely accessible AlphaFold Protein Structure Database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "Z4uj6blxUz3_",
   "metadata": {
    "cellView": "form",
    "id": "Z4uj6blxUz3_"
   },
   "outputs": [],
   "source": [
    "# @title Setup Cell üèóÔ∏è\n",
    "\n",
    "###############################################################################\n",
    "###############################################################################\n",
    "# @title Setup\n",
    "# import libraries\n",
    "import logging\n",
    "\n",
    "logger = logging.getLogger(\"ex02\")\n",
    "\n",
    "# Set logging configuration\n",
    "logging.basicConfig(level=logging.DEBUG)  # , format='%(levelname)s:%(message)s')\n",
    "logger.setLevel(logging.DEBUG)\n",
    "\n",
    "import os, sys, re\n",
    "import warnings\n",
    "\n",
    "warnings.simplefilter(action=\"ignore\", category=FutureWarning)\n",
    "\n",
    "if \"SETUP_DONE\" not in dir():\n",
    "    logger.info(\"üöß Setup environment\")\n",
    "    from IPython.utils import io\n",
    "    from IPython.display import HTML\n",
    "    import tensorflow as tf\n",
    "    import jax\n",
    "    import jax.numpy as jnp\n",
    "    import numpy as np\n",
    "    import matplotlib\n",
    "    from matplotlib import animation\n",
    "    import matplotlib.pyplot as plt\n",
    "    import tqdm.notebook\n",
    "\n",
    "    TQDM_BAR_FORMAT = (\n",
    "        \"{l_bar}{bar}| {n_fmt}/{total_fmt} [elapsed: {elapsed} remaining: {remaining}]\"\n",
    "    )\n",
    "\n",
    "    with io.capture_output() as captured:\n",
    "        if not os.path.isdir(\"af_backprop\"):\n",
    "            logger.info(\"üß© Downloading and installing af_backprop\")\n",
    "            %shell git clone https://github.com/sokrypton/af_backprop.git\n",
    "            %shell pip -q install biopython==1.79 dm-haiku ml-collections py3Dmol\n",
    "            %shell wget -qnc https://raw.githubusercontent.com/sokrypton/ColabFold/main/beta/colabfold.py\n",
    "        if not os.path.isdir(\"params\"):\n",
    "            logger.info(\"üß© Downloading and installing alphafold_parameters\")\n",
    "            %shell mkdir params\n",
    "            %shell apt-get install aria2 -qq\n",
    "            %shell aria2c -q -x 16 https://storage.googleapis.com/alphafold/alphafold_params_2021-07-14.tar\n",
    "            %shell tar -xf alphafold_params_2021-07-14.tar -C params\n",
    "            # %shell curl -fsSL https://storage.googleapis.com/alphafold/alphafold_params_2021-07-14.tar | tar x -C params\n",
    "\n",
    "    # configure which device to use\n",
    "    try:\n",
    "        # check if TPU is available\n",
    "        import jax.tools.colab_tpu\n",
    "\n",
    "        jax.tools.colab_tpu.setup_tpu()\n",
    "        logger.info(\"Running on TPU\")\n",
    "        DEVICE = \"tpu\"\n",
    "        logger.critical(\"‚ò†Ô∏è TPU is not supported 13.11.24\")\n",
    "        raise NotImplementedError(\"TPU is not supported\")\n",
    "    except:\n",
    "        if jax.local_devices()[0].platform == \"cpu\":\n",
    "            logger.warning(\"WARNING: no GPU detected, will be using CPU\")\n",
    "            DEVICE = \"cpu\"\n",
    "        else:\n",
    "            logger.info(\"Running on GPU\")\n",
    "            DEVICE = \"gpu\"\n",
    "            # disable GPU on tensorflow\n",
    "            tf.config.set_visible_devices([], \"GPU\")\n",
    "\n",
    "    # import libraries\n",
    "    sys.path.append(\"af_backprop\")\n",
    "\n",
    "    SETUP_DONE = True\n",
    "\n",
    "logger.info(\"‚úÖ Setup done\")\n",
    "\n",
    "if \"LIBRARY_IMPORTED\" not in dir():\n",
    "    logger.info(\"üöß Import libraries\")\n",
    "    from utils import update_seq, update_aatype, get_plddt, get_pae\n",
    "    import colabfold as cf\n",
    "    from alphafold.common import protein\n",
    "    from alphafold.data import pipeline\n",
    "    from alphafold.model import data, config, model\n",
    "    from alphafold.common import residue_constants\n",
    "\n",
    "    # custom functions\n",
    "    def clear_mem():\n",
    "        backend = jax.extend.backend.get_backend()\n",
    "        for buf in backend.live_buffers():\n",
    "            buf.delete()\n",
    "\n",
    "    def setup_model(max_len, model_name=\"model_3_ptm\"):\n",
    "        clear_mem()\n",
    "\n",
    "        # setup model\n",
    "        cfg = config.model_config(\"model_5_ptm\")\n",
    "        cfg.model.num_recycle = 0\n",
    "        cfg.data.common.num_recycle = 0\n",
    "        cfg.data.eval.max_msa_clusters = 1\n",
    "        cfg.data.common.max_extra_msa = 1\n",
    "        cfg.data.eval.masked_msa_replace_fraction = 0\n",
    "        cfg.model.global_config.subbatch_size = None\n",
    "        model_params = data.get_model_haiku_params(model_name=model_name, data_dir=\".\")\n",
    "        model_runner = model.RunModel(cfg, model_params, is_training=False)\n",
    "\n",
    "        seq = \"A\" * max_len\n",
    "        length = len(seq)\n",
    "        feature_dict = {\n",
    "            **pipeline.make_sequence_features(\n",
    "                sequence=seq, description=\"none\", num_res=length\n",
    "            ),\n",
    "            **pipeline.make_msa_features(\n",
    "                msas=[[seq]], deletion_matrices=[[[0] * length]]\n",
    "            ),\n",
    "        }\n",
    "        inputs = model_runner.process_features(feature_dict, random_seed=0)\n",
    "\n",
    "        def runner(I):\n",
    "            # update sequence\n",
    "            inputs = I[\"inputs\"]\n",
    "            inputs[\"prev\"] = I[\"prev\"]\n",
    "\n",
    "            seq = jax.nn.one_hot(I[\"seq\"], 20)\n",
    "            update_seq(seq, inputs)\n",
    "            update_aatype(inputs[\"target_feat\"][..., 1:], inputs)\n",
    "\n",
    "            # mask prediction\n",
    "            mask = jnp.arange(inputs[\"residue_index\"].shape[0]) < I[\"length\"]\n",
    "            inputs[\"seq_mask\"] = inputs[\"seq_mask\"].at[:].set(mask)\n",
    "            inputs[\"msa_mask\"] = inputs[\"msa_mask\"].at[:].set(mask)\n",
    "            inputs[\"residue_index\"] = jnp.where(mask, inputs[\"residue_index\"], 0)\n",
    "\n",
    "            # get prediction\n",
    "            key = jax.random.PRNGKey(0)\n",
    "            outputs = model_runner.apply(I[\"params\"], key, inputs)\n",
    "\n",
    "            aux = {\n",
    "                \"final_atom_positions\": outputs[\"structure_module\"][\n",
    "                    \"final_atom_positions\"\n",
    "                ],\n",
    "                \"final_atom_mask\": outputs[\"structure_module\"][\"final_atom_mask\"],\n",
    "                \"plddt\": get_plddt(outputs),\n",
    "                \"pae\": get_pae(outputs),\n",
    "                \"length\": I[\"length\"],\n",
    "                \"seq\": I[\"seq\"],\n",
    "                \"prev\": outputs[\"prev\"],\n",
    "                \"residue_idx\": inputs[\"residue_index\"][0],\n",
    "            }\n",
    "            return aux\n",
    "\n",
    "        return jax.jit(runner), {\n",
    "            \"inputs\": inputs,\n",
    "            \"params\": model_params,\n",
    "            \"length\": max_len,\n",
    "        }\n",
    "\n",
    "    def save_pdb(outs, filename):\n",
    "        \"\"\"save pdb coordinates\"\"\"\n",
    "        p = {\n",
    "            \"residue_index\": outs[\"residue_idx\"] + 1,\n",
    "            \"aatype\": outs[\"seq\"],\n",
    "            \"atom_positions\": outs[\"final_atom_positions\"],\n",
    "            \"atom_mask\": outs[\"final_atom_mask\"],\n",
    "            \"plddt\": outs[\"plddt\"],\n",
    "        }\n",
    "        p = jax.tree_util.tree_map(lambda x: x[: outs[\"length\"]], p)\n",
    "        b_factors = 100 * p.pop(\"plddt\")[:, None] * p[\"atom_mask\"]\n",
    "        p = protein.Protein(**p, b_factors=b_factors)\n",
    "        pdb_lines = protein.to_pdb(p)\n",
    "        with open(filename, \"w\") as f:\n",
    "            f.write(pdb_lines)\n",
    "\n",
    "    def make_animation(positions, plddts, Ls=None, line_w=2.0, dpi=100):\n",
    "        def ca_align_to_last(positions):\n",
    "            def align(P, Q):\n",
    "                p = P - P.mean(0, keepdims=True)\n",
    "                q = Q - Q.mean(0, keepdims=True)\n",
    "                return p @ cf.kabsch(p, q)\n",
    "\n",
    "            pos = positions[-1, :, 1, :] - positions[-1, :, 1, :].mean(0, keepdims=True)\n",
    "            best_2D_view = pos @ cf.kabsch(pos, pos, return_v=True)\n",
    "\n",
    "            new_positions = []\n",
    "            for i in range(len(positions)):\n",
    "                new_positions.append(align(positions[i, :, 1, :], best_2D_view))\n",
    "            return np.asarray(new_positions)\n",
    "\n",
    "        # align all to last recycle\n",
    "        pos = ca_align_to_last(positions)\n",
    "\n",
    "        fig, (ax1, ax2, ax3) = plt.subplots(1, 3)\n",
    "        fig.subplots_adjust(top=0.90, bottom=0.10, right=1, left=0, hspace=0, wspace=0)\n",
    "        fig.set_figwidth(13)\n",
    "        fig.set_figheight(5)\n",
    "        fig.set_dpi(dpi)\n",
    "\n",
    "        xy_min = pos[..., :2].min() - 1\n",
    "        xy_max = pos[..., :2].max() + 1\n",
    "\n",
    "        for ax in [ax1, ax3]:\n",
    "            ax.set_xlim(xy_min, xy_max)\n",
    "            ax.set_ylim(xy_min, xy_max)\n",
    "            ax.axis(False)\n",
    "\n",
    "        ax2.set_xlabel(\"positions\")\n",
    "        ax2.set_ylabel(\"pLDDT\")\n",
    "        ax2.set_ylim(0, 100)\n",
    "\n",
    "        ims = []\n",
    "        for k, (xyz, plddt) in enumerate(zip(pos, plddts)):\n",
    "            ims.append([])\n",
    "            im2 = ax2.plot(plddt, animated=True, color=\"black\")[0]\n",
    "            tt2 = cf.add_text(f\"recycle={k}\", ax2)\n",
    "            tt3 = cf.add_text(f\"pLDDT={plddt.mean():.3f}\", ax3)\n",
    "            if Ls is None or len(Ls) == 1:\n",
    "                tt1 = cf.add_text(\"colored by N->C\", ax1)\n",
    "                ims[-1] += [cf.plot_pseudo_3D(xyz, ax=ax1, line_w=line_w)]\n",
    "            else:\n",
    "                # color by chain\n",
    "                tt1 = cf.add_text(\"colored by chain\", ax1)\n",
    "                c = np.concatenate([[n] * L for n, L in enumerate(Ls)])\n",
    "                ims[-1] += [\n",
    "                    cf.plot_pseudo_3D(\n",
    "                        xyz,\n",
    "                        c=c,\n",
    "                        cmap=cf.pymol_cmap,\n",
    "                        cmin=0,\n",
    "                        cmax=39,\n",
    "                        line_w=line_w,\n",
    "                        ax=ax1,\n",
    "                    )\n",
    "                ]\n",
    "\n",
    "            ims[-1] += [im2, tt1, tt2, tt3]\n",
    "            ims[-1] += [\n",
    "                cf.plot_pseudo_3D(xyz, c=plddt, cmin=50, cmax=90, ax=ax3, line_w=line_w)\n",
    "            ]\n",
    "\n",
    "        ani = animation.ArtistAnimation(fig, ims, blit=True, interval=120)\n",
    "        plt.close()\n",
    "        return ani.to_html5_video()\n",
    "\n",
    "    LIBRARY_IMPORTED = True\n",
    "\n",
    "logger.info(\"‚úÖ Libraries imported\")\n",
    "\n",
    "\n",
    "# Cell 2: Define Prediction Function\n",
    "def predict_structure(\n",
    "    sequence,\n",
    "    recycles=48,\n",
    "    color=\"confidence\",\n",
    "    show_sidechains=True,\n",
    "    show_mainchains=False,\n",
    "):\n",
    "    # initialize\n",
    "    # if \"current_seq\" not in dir():\n",
    "    current_seq = \"\"\n",
    "    r = -1\n",
    "    max_length = -1\n",
    "\n",
    "    # collect user inputs\n",
    "    ori_sequence = re.sub(\"[^A-Z\\/\\:]\", \"\", sequence.upper())\n",
    "\n",
    "    # Check if the sequence is empty after cleaning\n",
    "    if not ori_sequence:\n",
    "        logger.critical(\n",
    "            \"‚ùå Input sequence is empty or invalid. Please enter a valid amino acid sequence.\".upper()\n",
    "        )\n",
    "        return None\n",
    "\n",
    "    Ls = [len(s) for s in ori_sequence.replace(\":\", \"/\").split(\"/\")]\n",
    "    sequence = re.sub(\"[^A-Z]\", \"\", ori_sequence)\n",
    "    length = len(sequence)\n",
    "\n",
    "    # avoid recompiling if length within 25\n",
    "    if length > max_length or (max_length - length) > 25:\n",
    "        max_length = length + 25\n",
    "        runner, I = setup_model(max_length)\n",
    "\n",
    "    if ori_sequence != current_seq:\n",
    "        outs = []\n",
    "        positions = []\n",
    "        plddts = []\n",
    "        paes = []\n",
    "        r = -1\n",
    "\n",
    "        # pad sequence to max length\n",
    "        seq = np.array([residue_constants.restype_order.get(aa, 0) for aa in sequence])\n",
    "        seq = np.pad(seq, [0, max_length - length], constant_values=-1)\n",
    "\n",
    "        # update inputs, restart recycle\n",
    "        I.update(\n",
    "            {\n",
    "                \"seq\": seq,\n",
    "                \"length\": length,\n",
    "                \"prev\": {\n",
    "                    \"prev_msa_first_row\": np.zeros([max_length, 256]),\n",
    "                    \"prev_pair\": np.zeros([max_length, max_length, 128]),\n",
    "                    \"prev_pos\": np.zeros([max_length, 37, 3]),\n",
    "                },\n",
    "            }\n",
    "        )\n",
    "\n",
    "        I[\"inputs\"][\"residue_index\"][:] = cf.chain_break(\n",
    "            np.arange(max_length), Ls, length=32\n",
    "        )\n",
    "        current_seq = ori_sequence\n",
    "\n",
    "    # run for defined number of recycles\n",
    "    with tqdm.notebook.tqdm(total=(recycles + 1), bar_format=TQDM_BAR_FORMAT) as pbar:\n",
    "        p = 0\n",
    "        while p < min(r + 1, recycles + 1):\n",
    "            pbar.update(1)\n",
    "            p += 1\n",
    "        while r < recycles:\n",
    "            O = runner(I)\n",
    "            O = jax.tree_util.tree_map(lambda x: np.asarray(x), O)\n",
    "            positions.append(O[\"final_atom_positions\"][:length])\n",
    "            plddts.append(O[\"plddt\"][:length])\n",
    "            paes.append(O[\"pae\"][:length, :length])\n",
    "            I[\"prev\"] = O[\"prev\"]\n",
    "            outs.append(O)\n",
    "            r += 1\n",
    "            pbar.update(1)\n",
    "\n",
    "    if color == \"confidence\":\n",
    "        color = \"lDDT\"\n",
    "\n",
    "    print(f\"plotting prediction at recycle={recycles}\")\n",
    "    save_pdb(outs[recycles], \"out.pdb\")\n",
    "    v = cf.show_pdb(\n",
    "        \"out.pdb\",\n",
    "        show_sidechains,\n",
    "        show_mainchains,\n",
    "        color,\n",
    "        color_HP=True,\n",
    "        size=(800, 480),\n",
    "        Ls=Ls,\n",
    "    )\n",
    "    v.setHoverable(\n",
    "        {},\n",
    "        True,\n",
    "        \"\"\"function(atom,viewer,event,container){if(!atom.label){atom.label=viewer.addLabel(\"      \"+atom.resn+\":\"+atom.resi,{position:atom,backgroundColor:'mintcream',fontColor:'black'});}}\"\"\",\n",
    "        \"\"\"function(atom,viewer){if(atom.label){viewer.removeLabel(atom.label);delete atom.label;}}\"\"\",\n",
    "    )\n",
    "    v.show()\n",
    "    if color == \"lDDT\":\n",
    "        cf.plot_plddt_legend().show()\n",
    "\n",
    "    # add confidence plots\n",
    "    cf.plot_confidence(plddts[recycles] * 100, paes[recycles], Ls=Ls).show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "Ur6jDLmffJ_3",
   "metadata": {
    "id": "Ur6jDLmffJ_3"
   },
   "source": [
    "## Alphafold_single Tasks\n",
    "\n",
    "This tutorial is based on [Alphafold_single](https://colab.research.google.com/github/sokrypton/af_backprop/blob/beta/examples/AlphaFold_single.ipynb). Follow the instructions:\n",
    "\n",
    "- Patience... The first time you run the cell below it will take 1 minitue to setup, after that it should run in seconds (after each change).\n",
    "- click the little ‚ñ∂ play icon to the left of each cell below.\n",
    "- For 3D display, hold mouseover aminoacid to get name and position number\n",
    "- use \"/\" to specify chainbreaks, (eg. sequence=\"AAA/AAA\")\n",
    "\n",
    "Tasks:\n",
    "\n",
    " 1. Change the sequence of aminoacids to a different one (e.g. all Alanines, all Leucines, etc.). What is the difference in the predicted structure?\n",
    "\n",
    " 1. Design a sequence that fold into a Helix (see figure)\n",
    "\n",
    " 2. Design a sequence that fold into Two Helices (see figure)\n",
    "\n",
    " 3. Design a sequence that fold into Beta Sheets (see figure)\n",
    "\n",
    "<img src=\"https://github.com/yerkoescalona/structural_bioinformatics/blob/w2023/ex02/challenge1.png?raw=1\" alt=\"Drawing\"/>\n",
    "\n",
    " 5. Design a sequence that fold into 4 helix bundle (take a look at e.g. 3VJF)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    " **Hint**: Use the following information to design your sequences\n",
    "\n",
    "<img src=\"https://github.com/yerkoescalona/structural_bioinformatics/blob/w2023/ex02/statistics.png?raw=1\" alt=\"Drawing\"/>\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bYCIEttaT9Tt",
   "metadata": {
    "cellView": "form",
    "id": "bYCIEttaT9Tt"
   },
   "outputs": [],
   "source": [
    "# Cell 3: Predict Structure for a Sequence\n",
    "\n",
    "# @title Change the sequence of aminoacids to a different one (e.g. all Alanines, all Leucines, etc.). What is the difference in the predicted structure?\n",
    "sequence = \"GGGGGGGGGG\"  # @param {type:\"string\"}\n",
    "recycles = 48  # @param [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\", \"24\", \"48\"] {type:\"raw\"}\n",
    "color = \"rainbow\"  # @param [\"chain\", \"confidence\", \"rainbow\"]\n",
    "show_sidechains = True  # @param {type:\"boolean\"}\n",
    "show_mainchains = True  # @param {type:\"boolean\"}\n",
    "\n",
    "predict_structure(\n",
    "    sequence=sequence,\n",
    "    recycles=recycles,\n",
    "    color=color,\n",
    "    show_sidechains=show_sidechains,\n",
    "    show_mainchains=show_mainchains,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "jVcT_sNAZnyR",
   "metadata": {
    "cellView": "form",
    "id": "jVcT_sNAZnyR"
   },
   "outputs": [],
   "source": [
    "# Cell 3: Predict Structure for a Sequence\n",
    "\n",
    "# @title Design sequences that fold into a ALPHA HELIX secondary structures (see figure above)\n",
    "sequence = \"\"  # @param {type:\"string\"}\n",
    "recycles = 48  # @param [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\", \"24\", \"48\"] {type:\"raw\"}\n",
    "color = \"rainbow\"  # @param [\"chain\", \"confidence\", \"rainbow\"]\n",
    "show_sidechains = True  # @param {type:\"boolean\"}\n",
    "show_mainchains = True  # @param {type:\"boolean\"}\n",
    "\n",
    "predict_structure(\n",
    "    sequence,\n",
    "    recycles=recycles,\n",
    "    color=color,\n",
    "    show_sidechains=show_sidechains,\n",
    "    show_mainchains=show_mainchains,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eMX7Tyk_xGCS",
   "metadata": {
    "cellView": "form",
    "id": "eMX7Tyk_xGCS"
   },
   "outputs": [],
   "source": [
    "# Cell 3: Predict Structure for a Sequence\n",
    "\n",
    "# @title Design sequences that fold into TWO HELICES (see figure above)\n",
    "sequence = \"\"  # @param {type:\"string\"}\n",
    "recycles = 48  # @param [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\", \"24\", \"48\"] {type:\"raw\"}\n",
    "color = \"rainbow\"  # @param [\"chain\", \"confidence\", \"rainbow\"]\n",
    "show_sidechains = True  # @param {type:\"boolean\"}\n",
    "show_mainchains = True  # @param {type:\"boolean\"}\n",
    "\n",
    "predict_structure(\n",
    "    sequence,\n",
    "    recycles=recycles,\n",
    "    color=color,\n",
    "    show_sidechains=show_sidechains,\n",
    "    show_mainchains=show_mainchains,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "_WRUMdbqxPjJ",
   "metadata": {
    "cellView": "form",
    "id": "_WRUMdbqxPjJ"
   },
   "outputs": [],
   "source": [
    "# Cell 3: Predict Structure for a Sequence\n",
    "\n",
    "# @title Design sequences that fold into BETA SHEETS (see figure above)\n",
    "sequence = \"\"  # @param {type:\"string\"}\n",
    "recycles = 48  # @param [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\", \"24\", \"48\"] {type:\"raw\"}\n",
    "color = \"rainbow\"  # @param [\"chain\", \"confidence\", \"rainbow\"]\n",
    "show_sidechains = True  # @param {type:\"boolean\"}\n",
    "show_mainchains = True  # @param {type:\"boolean\"}\n",
    "\n",
    "predict_structure(\n",
    "    sequence,\n",
    "    recycles=recycles,\n",
    "    color=color,\n",
    "    show_sidechains=show_sidechains,\n",
    "    show_mainchains=show_mainchains,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "vc0JrCVDZxO2",
   "metadata": {
    "cellView": "form",
    "id": "vc0JrCVDZxO2"
   },
   "outputs": [],
   "source": [
    "# Cell 3: Predict Structure for a Sequence\n",
    "\n",
    "# @title Design a sequence that fold into 4 helix bundle (take a look at e.g. 3VJF)\n",
    "sequence = \"\"  # @param {type:\"string\"}\n",
    "recycles = 48  # @param [\"0\", \"1\", \"2\", \"3\", \"6\", \"12\", \"24\", \"48\"] {type:\"raw\"}\n",
    "color = \"rainbow\"  # @param [\"chain\", \"confidence\", \"rainbow\"]\n",
    "show_sidechains = True  # @param {type:\"boolean\"}\n",
    "show_mainchains = True  # @param {type:\"boolean\"}\n",
    "\n",
    "predict_structure(\n",
    "    sequence,\n",
    "    recycles=recycles,\n",
    "    color=color,\n",
    "    show_sidechains=show_sidechains,\n",
    "    show_mainchains=show_mainchains,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e0c5586-affa-4c98-b2d3-2375215a89ce",
   "metadata": {
    "collapsed": false,
    "id": "0e0c5586-affa-4c98-b2d3-2375215a89ce"
   },
   "source": [
    "## For the project\n",
    "\n",
    "Let's put into practice what we have learned so far.\n",
    "We will use Alphafold to predict the structure of a protein and then we will try to disrupt it by introducing mutations.\n",
    "\n",
    "Tasks:\n",
    "1. Go to the official [Alphafold2](https://colab.research.google.com/github/deepmind/alphafold/blob/main/notebooks/AlphaFold.ipynb) notebook or [ColabFold](https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb), read the instructions and predict the structure of your protein its sequence as an input. **Important Note**: There is a high probability that your protein is already in the database. Look in the [AlphaFold Protein Structure Database](https://alphafold.ebi.ac.uk/).  If so, download the model and respond to the following questions.\n",
    "\n",
    "    Questions:\n",
    "    - what is the confidence score of your prediction?\n",
    "    - what are the most confident regions?\n",
    "    - align your crystal structure with the predicted one. What is the RMSD? (To calculate the RMSD between two proteins, you can use PyMol, load the two structures and use the command `rms_cur protein1 protein2` to get the value)\n",
    "\n",
    "2. Use Alphafold or ColabFold to find a (minimum) set of mutations to disrupt your protein.\n",
    "\n",
    "    Questions:\n",
    "    - what are the most common disruptive mutations?\n",
    "    - are there subsets in your list of mutations? Disrupting different regions or functions?\n",
    "\n",
    "\n",
    "\n",
    "**Hint**: consider the following table of stabilizing forces in proteins\n",
    "\n",
    " <img src=\"https://github.com/yerkoescalona/structural_bioinformatics/blob/w2023/ex02/structure_stabilizing_forces.png?raw=1\" alt=\"Drawing\"/>\n",
    "\n",
    "**Hint 2**: Use your knowledge to analyze a multiple sequence alignment (MSA). You can use [Clustal Omega](https://toolkit.tuebingen.mpg.de/tools/clustalo) or [MAFFT](https://toolkit.tuebingen.mpg.de/tools/mafft).\n",
    "\n",
    "Go [Here](https://colab.research.google.com/github/yerkoescalona/structural_bioinformatics/blob/main/ex02/msa.ipynb) for an example. With this notebook you can perform a multiple sequence alignment on a series of FASTA sequences (FtsZ sequences as example).\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3cfbad639a217f6",
   "metadata": {
    "collapsed": false,
    "id": "d3cfbad639a217f6"
   },
   "source": [
    "Happy modeling!"
   ]
  }
 ],
 "metadata": {
  "accelerator": "GPU",
  "colab": {
   "gpuType": "T4",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
